dali/pipeline/operator/builtin/external_source.h

```diff
@@ -506,6 +508,23 @@ class ExternalSource : public Operator<Backend>, virtual public BatchSizeProvide
                   " and the current type is ", batch.type_info().name(), "."));
     previous_dtype_ = batch.type();
 
+    auto current_ndim = batch.shape().sample_dim();
+    DALI_ENFORCE(ndim_ == -1 || ndim_ == current_ndim,
+      make_string("ExternalSource expected data with ", ndim_, " dimensions and got ",
+                   current_ndim, " dimensions."));
+
+    DALI_ENFORCE(previous_ndim_ == -1 || previous_ndim_ == current_ndim,
```

Comment:
Technically, we could use `ndim_` for previous and select the error message based `spec_.HasArgument("ndim")`

Reply:
done